### PR TITLE
Add OpenTelemetry tracing to 'compute_tools'

### PR DESCRIPTION
**Pull request by hlinnaka**
This consists of 3 individual commits:

1. Configure OTLP exporter that exports the traces to a running collector service. By default, it will try to connect to 'http://localhost:4317', but it can be changed with the 'OTEL_EXPORTER_OTLP_ENDPOINT` env variable

2. Extract span context from the compute spec, for the startup actions. All the startup operations, like starting up Postgres and changing the roles and databases, are performed in that context. The cloud console passes the context, so that the startup operations become part of the same trace as the operations in the cloud console. This allows getting a trace of the whole compute startup process that covers all the actions in the cloud console and in 'compute_tools'. In the future, we could pass the context all the way from the proxy, but that's not done here yet.

3. Extract span context from incoming HTTP requests. This allows tracing the actions done e.g. for `/check_writability` together with the process that send the request.

I believe the default configuration for the OTLP exporter is to export a full trace of *all* operations. That might be too much in production, although 'compute_tools' itself won't produce much trace and the compute startup is a very heavy-weight operation in comparison. But if we add more tracing support in the future, we probably want to configure it more carefully in production

### CODE REVIEW

**Comment by hlinnaka** [comment]
I reworked this quite a bit, compared to the original version. Please read the individual commit messages for details.

File: workspace_hack/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,9 @@ clap = { version = "4", features = ["derive", "string"] }
 crossbeam-utils = { version = "0.8" }
 either = { version = "1" }
 fail = { version = "0.5", default-features = false, features = ["failpoints"] }
+futures = { version = "0.3" }
 futures-channel = { version = "0.3", features = ["sink"] }
+futures-executor = { version = "0.3" }
```
Comment (koivunej) [inline workspace_hack/Cargo.toml:25]: Why is this required? Cannot see it being used..

File: libs/tracing-utils/src/http.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        otel.name= %otel_name,
+        http.method = %req.method(),
+        http.status_code = tracing::field::Empty,
+    );
+    let parent_ctx = extract_remote_context(req.headers());
+    span.set_parent(parent_ctx);
+    let _ = span.enter();
+
+    // Handle the request
+    let response = handler(req).await;
```
Comment (koivunej) [inline libs/tracing-utils/src/http.rs:67]: This seems wrong, instead with futures the instrument wrapping future to enter/exit with each poll should be the way to go:

```suggestion
    // Handle the request
    let response = handler(req).await.instrument(span);
```

File: workspace_hack/Cargo.toml (hunk 1 above)
Comment (hlinnaka) [reply workspace_hack/Cargo.toml:25]: Hmm, `cargo hakari generate` wants to put it there..

To see where it's used, I temporarily removed it from workspace_hack, and ran `cargo tree`. It's used in two places now:

* (several packages) -> futures -> futures-executor
* comute_tools -> opentelemetry -> opentelemetry_sdk -> futures-executor

The second dependency to it is new, so I guess "hakari" wants to add it to workspace_hack now because there are now two dependency paths to it.

File: libs/tracing-utils/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+///
+/// However, we only support a subset of those options:
+///
+/// - OTEL_SDK_DISABLED is supported. The default is "false", meaning tracing
+///   is enabled by default. Set it to "true" to disable.
+///
+/// - We use the OTLP exporter, with HTTP protocol. Most of the OTEL_EXPORTER_OTLP_*
+///   settings specified in
+///   <https://opentelemetry.io/docs/reference/specification/protocol/exporter/>
+///   are supported, asthey are handled by the `opentelemetry-otlp` crate.
```
Comment (koivunej) [inline libs/tracing-utils/src/lib.rs:62]:
```suggestion
///   are supported, as they are handled by the `opentelemetry-otlp` crate.
```

(Finally reading the comments nearing towards the end of review :))

File: libs/tracing-utils/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_otlp::{OTEL_EXPORTER_OTLP_ENDPOINT, OTEL_EXPORTER_OTLP_TRACES_ENDPOINT};
+
+pub use tracing_opentelemetry::OpenTelemetryLayer;
+
+pub mod http;
+
+///
+/// Set up OpenTelemetry exporter, using configuration from environment variables.
+/// `service_name` is set as the OpenTelemetry 'service.name' resource (see
```
Comment (koivunej) [inline libs/tracing-utils/src/lib.rs:48]:
```suggestion
/// Set up OpenTelemetry exporter, using configuration from environment variables.
///
/// `service_name` is set as the OpenTelemetry 'service.name' resource (see
```

rustdoc uses the first line as "summary" in listings, similar to git commit messages have the subject line.

File: libs/tracing-utils/src/http.rs (hunk 2 above)
Comment (koivunej) [reply libs/tracing-utils/src/http.rs:67]: Also, should we catch_unwind here?

File: workspace_hack/Cargo.toml [hunk 5]
```diff
@@ -36,12 +38,16 @@ prost = { version = "0.11" }
 rand = { version = "0.8", features = ["small_rng"] }
 regex = { version = "1" }
 regex-syntax = { version = "0.6" }
+reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "rustls-tls"] }
+ring = { version = "0.16", features = ["std"] }
+rustls = { version = "0.20", features = ["dangerous_configuration"] }
```
Comment (koivunej) [inline workspace_hack/Cargo.toml:43]: `dangerous_configuration` comes from reqwest, otlp doesn't set it directly.

File: workspace_hack/Cargo.toml [hunk 6]
```diff
@@ -36,12 +38,16 @@ prost = { version = "0.11" }
 rand = { version = "0.8", features = ["small_rng"] }
 regex = { version = "1" }
 regex-syntax = { version = "0.6" }
+reqwest = { version = "0.11", default-features = false, features = ["blocking", "json", "rustls-tls"] }
```
Comment (koivunej) [inline workspace_hack/Cargo.toml:41]: surprised to see the blocking feature here. Even though it's only a bit more methods enabled, I cannot immediatedly see why it's being enabled.

File: workspace_hack/Cargo.toml (hunk 1 above)
Comment (koivunej) [reply workspace_hack/Cargo.toml:25]: You are correct. Great idea disabling the workspace hack and then querying. Very odd to see it being depended on directly: https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/Cargo.toml#L21 and used directly from https://github.com/open-telemetry/opentelemetry-rust/blob/0ef87ad78a6b83a91c14a7926efae101a8b111c2/opentelemetry-sdk/src/metrics/controllers/basic.rs#L174. Maybe there's a good reason.

Resolving...

File: libs/tracing-utils/src/http.rs (hunk 2 above)
Comment (koivunej) [reply libs/tracing-utils/src/http.rs]: I see you changed this. I'd double check with jaeger that the span cloning doesn't create any surprising effects on the traces..

File: workspace_hack/Cargo.toml (hunk 5 above)
Comment (koivunej) [reply workspace_hack/Cargo.toml:43]: We cannot add comments to this file, so just resolving.

File: compute_tools/src/bin/compute_ctl.rs [hunk 7]
```diff
@@ -84,6 +84,28 @@ fn main() -> Result<()> {
         }
     };
 
+    // Extract OpenTelemetry context for the startup actions from the spec.
+    //
+    // This is used to propagate the context for the 'start_compute' operation
+    // from the neon control plane. This allows linking together the wider
+    // 'start_compute' operation that creates the compute container, with the
```
Comment (koivunej) [inline compute_tools/src/bin/compute_ctl.rs:91]: Stray "with the".

File: libs/tracing-utils/src/http.rs (hunk 2 above)
Comment (hlinnaka) [reply libs/tracing-utils/src/http.rs]: Yep, tested it.

`Span::current()` also returns a clone of the Span, rather than `&Span`. I was surprised. A Span is apparently just a reference to the underlying "span thingie", like an Arc. Although looking at the implementation of `Span::clone`, it's a little more complicated than incrementing a reference count.

Comment (hlinnaka) [reply libs/tracing-utils/src/http.rs]: Good catch on the original issue, thanks!

File: compute_tools/src/bin/compute_ctl.rs (hunk 7 above)
Comment (hlinnaka) [reply compute_tools/src/bin/compute_ctl.rs:92]: Fixed, and improved the comment a little in general.

**Review by koivunej — APPROVED** [review]
Good catch with the `set_text_map_propagator`, I was going to ask about the extractor usage on the spec, should could it just be a normal span but perhaps that'd be too much of a burden for normal development logging nor am I sure if you can just unpack a `HashMap<String, String>` as span args.

All points I raised seemed to be fixed. Wondering if @funbringer will have a chance to peek as well?

**Comment by funbringer** [comment]
Sorry, right now I'm out of capacity; can't give this a proper review.
